pandas/core/missing.py

```diff
@@ -314,6 +287,73 @@ def interpolate_1d(
         return result
 
 
+def _derive_indices_of_nans_to_preserve(
+    yvalues: ArrayLike,
+    limit: Optional[int] = None,
+    limit_area: Optional[str] = None,
+    limit_direction: Optional[str] = None,
+):
+    """
+    Derive the indices of NaNs that shall be preserved after interpolation
+    This function is called by `interpolate_1d` and takes the arguments with
+    the same name from there. In `interpolate_1d`, after performing the
+    interpolation, the list of indices of NaNs to preserve is used to put
+    NaNs in the desired locations.
+
+    Parameters
+    ----------
+    yvalues: ArrayLike
+        1-d array of values of the initial Series or DataFrame
+    limit: int
+    limit_area: str
+    limit_direction: str
+
+    Returns
+    -------
+    preserve_nans: set
+        Set of index pointers to where NaNs should be preserved in `yvalues`
+    """
+
+    invalid = isna(yvalues)
+    valid = ~invalid
+
+    # These are sets of index pointers to invalid values... i.e. {0, 1, etc...
+    all_nans = set(np.flatnonzero(invalid))
+    start_nans = set(range(find_valid_index(yvalues, "first")))
+    end_nans = set(range(1 + find_valid_index(yvalues, "last"), len(valid)))
+    mid_nans = all_nans - start_nans - end_nans
+
+    # Like the sets above, preserve_nans contains indices of invalid values,
+    # but in this case, it is the final set of indices that need to be
+    # preserved as NaN after the interpolation.
+
+    # For example if limit_direction='forward' then preserve_nans will
+    # contain indices of NaNs at the beginning of the series, and NaNs that
+    # are more than'limit' away from the prior non-NaN.
+
+    # set preserve_nans based on direction using _interp_limit
+    if limit_direction == "forward":
+        preserve_nans = start_nans | set(_interp_limit(invalid, limit, 0))
+    elif limit_direction == "backward":
+        preserve_nans = end_nans | set(_interp_limit(invalid, 0, limit))
+    else:
+        # both directions... just use _interp_limit
+        preserve_nans = set(_interp_limit(invalid, limit, limit))
+
+    # if limit_area is set, add either mid or outside indices
+    # to preserve_nans GH #16284
+    if limit_area == "inside":
+        # preserve NaNs on the outside
+        preserve_nans |= start_nans | end_nans
+    elif limit_area == "outside":
+        # preserve NaNs on the inside
+        preserve_nans |= mid_nans
+
+    # sort preserve_nans and covert to list
+    preserve_nans = sorted(preserve_nans)
```

Comment:
@jreback Should I force a set here with
```
preserve_nans = set(sorted(preserve_nans))
```
so that the CI type validation does not complain?

Reply:
If you don't re-use the same variable should be OK, so something like `nans_list = sorted(preserve_nans)` should work

Reply:
Solved